File: lib/ParticleFilter/particle.cpp
#include "particle.h"

#include <random>

Particle Particle::createParticleInCell(int ID, double weight, Cell &cell)
{
    // Generate x and y coordinates randomly and never on the borders of the cell:
    int xCoordinate = cell.startX + 1 + rand() % (cell.stopX - cell.startX);
    int yCoordinate = cell.startY + 1 + rand() % (cell.stopY - cell.startY);
    int direction = rand() % 360; // Will create direction between 0-360

    return Particle(ID, xCoordinate, yCoordinate, direction, weight);
}

Particle::Particle()
{
}

Particle::Particle(int ID, int xCoordinate, int yCoordinate, int direction, double weight)
{
    this->ID = ID;
    this->xCoordinate = xCoordinate;
    this->yCoordinate = yCoordinate;
    this->direction = direction;
    this->weight = weight;
}

int Particle::getXCoordinate()
{
    return this->xCoordinate;
}

int Particle::getYcoordinate()
{
    return this->yCoordinate;
}

double Particle::getWeight() {
    return this->weight;
}

/// @brief Update the coordinates of the particle.
/// @param newXcoordinate New X coordinate of the particle.
/// @param newYCoordinate New Y coordinate of the particle.
void Particle::updateCoordinates(int newXcoordinate, int newYCoordinate)
{
    this->xCoordinate = newXcoordinate;
    this->yCoordinate = newYCoordinate;
}

/// @brief Update the weight of the particle.
/// @param newWeight New weight of the particle.
void Particle::updateWeight(double newWeight)
{
    this->weight = newWeight;
}

File: Python/LocalizationTable/Particle.py
import random
from Map.Cell import Cell


class Particle:
    @staticmethod
    def create_particle_in_cell(ID: int, weight: float, cell: "Cell"):
        # Generate random x and y coordinates within the cell bounds
        x_coordinate = random.randint(cell.start_x + 1, cell.stop_x - 1)
        y_coordinate = random.randint(cell.start_y + 1, cell.stop_y - 1)

        if not cell.contains_point(x_coordinate, y_coordinate):
            test = 10

        # Generate a random direction between 0 and 359 degrees
        direction = random.randint(0, 359)

        return Particle(ID, x_coordinate, y_coordinate, direction, weight, cell.id)

    def __init__(self, ID: int, x_coordinate: int, y_coordinate: int, direction: int, weight: float, cell_id:int):
        self.ID = ID
        self.x_coordinate = x_coordinate
        self.y_coordinate = y_coordinate
        self.cell_id = cell_id
        self.direction = direction
        self.weight = weight

    def get_x_coordinate(self) -> int:
        return self.x_coordinate

    def get_y_coordinate(self) -> int:
        return self.y_coordinate

    def get_weight(self) -> float:
        return self.weight

    def update_coordinates(self, new_x_coordinate: int, new_y_coordinate: int) -> None:
        self.x_coordinate = new_x_coordinate
        self.y_coordinate = new_y_coordinate

    def update_weight(self, new_weight: float) -> None:
        self.weight = new_weight
